### PR TITLE
Add replay protection to JSON-RPC

### PR DESCRIPTION
**Pull request by Mrtenz**
A nonce is stored in the application for each public key, and must be incremented by one for each JSON-RPC request.

### CODE REVIEW

File: packages/signer/src/api/ws.sagas.ts [hunk 1]
```diff
@@ -162,22 +184,34 @@ export function* handleRequest({ socket, request, data }: WebSocketMessage) {
...
+          error: { code: '-32600', message: 'Invalid request' }
+        })
+      )
+    );
+  }
+
+  // Dont serve requests with invalid nonces
+  const isValidNonce: boolean = yield call(verifyRequestNonce, jsonRpcRequest, publicKey);
+  if (!isValidNonce) {
+    return socket.send(
```
Comment (FrederikBolding) [inline packages/signer/src/api/ws.sagas.ts:210]: Can we test this?

File: packages/signer/src/api/ws.sagas.ts [hunk 2]
```diff
@@ -146,6 +149,25 @@ const getOriginHost = (request: IncomingMessage) => {
   }
 };
 
+/**
+ * Verifies the JSON-RPC request nonce based on the ID and public key. If the nonce is valid, the
```
Comment (FrederikBolding) [inline packages/signer/src/api/ws.sagas.ts:153]: Perhaps add something about: "we expect the JSON RPC ids to be nonces and increment for each request..."?

**Review by FrederikBolding — CHANGES_REQUESTED** [review]

File: packages/signer/src/api/ws.persist.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import type { PersistConfig } from '@signer/common';
+import { createPersistReducer, incrementNonce, wsSlice } from '@signer/common';
+
+const persistConfig: PersistConfig = {
+  key: wsSlice.name,
+  whitelistedActions: [incrementNonce.type],
+  whitelistedKeys: []
```
[the remark this answers is not in the document]
Comment (Mrtenz) [reply packages/signer/src/api/ws.persist.ts:7]: Yes, I initially used
```ts
{
  [publicKey: string]: number;
}
```
as type for the slice, but then realised I couldn't whitelist the keys. Changed it to have a `nonces` key but forgot to add it to the array here...

**Review by FrederikBolding — APPROVED** [review]
lgtm

**Comment by Mrtenz** [comment]
[sc-8974]

**Pull request merged**